crates/core/app/tests/common/test_node_builder_ext.rs

```diff
@@ -21,75 +23,59 @@ impl BuilderExt for Builder {
+            // Generate a penumbra validator with this consensus key, and a corresponding
+            // allocation of delegation tokens.
+            let (validator, allocation, _) = generate_penumbra_validator(consensus_vk);
+
+            // Add the validator to the staking component's genesis content.
+            trace!(?validator, "adding validator to staking genesis content");
+            content.stake_content.validators.push(validator);
+
+            // Add an allocation of delegation tokens to the shielded pool content.
+            trace!(
+                ?allocation,
+                "adding allocation to shielded pool genesis content"
+            );
+            content.shielded_pool_content.allocations.push(allocation);
+        }
+
         // Serialize the app state into bytes, and add it to the builder.
+        let app_state = AppState::Content(content);
         serde_json::to_vec(&app_state)
             .map_err(Self::Error::from)
             .map(|s| self.app_state(s))
     }
 }
 
-/// Injects the given collection of [`Validator`s][PenumbraValidator] into the app state.
-fn inject_penumbra_validators(
-    app_state: AppState,
-    validators: Vec<PenumbraValidator>,
-) -> Result<AppState, anyhow::Error> {
-    use AppState::{Checkpoint, Content};
-    match app_state {
-        Checkpoint(_) => anyhow::bail!("checkpoint app state isn't supported"),
-        Content(mut content) => {
-            // Inject the builder's validators into the staking component's genesis state...
-            let overwritten = std::mem::replace(&mut content.stake_content.validators, validators);
-            // ...and log a warning if this overwrote any validators already in the app state.
-            if !overwritten.is_empty() {
-                tracing::warn!(
-                    ?overwritten,
-                    "`with_penumbra_auto_app_state` overwrote validators in the given AppState"
-                )
-            }
-            Ok(Content(content))
-        }
-    }
-}
-
 /// Generates a [`Validator`][PenumbraValidator] given a consensus verification key.
-fn generate_penumbra_validator(
-    consensus_key: ed25519_consensus::VerificationKey,
-) -> PenumbraValidator {
-    /// A temporary stub for validator keys.
-    ///
-    /// An invalid key is intentionally provided here, until we have test coverage exercising the
-    /// use of these keys. Once we need it we will:
-    /// - generate a random signing key
-    /// - get its verification key
-    /// - use that for the identity key
-    /// - throw the signing key away
-    ///
-    /// NB: for now, we will use the same key for governance. See the documentation of
-    /// `GovernanceKey` for more information about cold storage of validator keys.
-    const INVALID_KEY_BYTES: [u8; 32] = [0; 32];
+pub fn generate_penumbra_validator(
+    consensus_key: &ed25519_consensus::VerificationKey,
+) -> (PenumbraValidator, Allocation, penumbra_keys::keys::SpendKey) {
```

Comment:
this should remain a private API, and not return the spend key

Reply:
fixed in ceb5d62

Reply:
3ca8c462e removes the spend key from the return value.